plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1093,50 +1098,253 @@ \subsection{Type erasure}
+proceeds to evaluate $M$ in the environment $\rho^\prime$: when any
+occurrences of $x$ are encountered, the machine retrieves the value
+$V$ from the environment and proceeds with evaluation as normal.  It
+is necessary to use closures because values may now contain variable
+names, and we have to know what the variables were bound to at the
+time when the value was created.
+
+\subsubsection{Discharging closures}
+Given a closure $(V,\rho)$ we denote by $\discharge{(V,\rho)}$ the
+result of \textit{discharging} the closure, ie, recursively replacing
+all of the variables occurring in $V$ by their bindings in
+$\rho$. Note that variables are bound to closures which may themselves
+need to be discharged: hence the use of the word ``recursively''. It
+can be shown that when any closure produced by the CEK machine is
+discharged it yields a value.
+
+\subsection{Definition of the CEK machine}
+The CEK machine is described in Figure~\ref{fig:cek-machine}.  It is
+  is very similar to the CK machine of Figure~\ref{fig:typed-ck-machine},
+  except that
+\begin{itemize}
+\item Every rule now involves environments
+\item A new rule has been added to deal with the evaluation of variables
+\item The rule involving lambda application has changed to replace substitution with
+the use of an extended environment
+\item When the machine terminates without error it now
+returns a closure $(V,\rho)$ rather than a value.  The invoker
+of the machine may wish to discharge this closure to obtain a value.
+\end{itemize}
+\noindent The major changes are highlighted in \blue{blue}. To evaluate a program
+\texttt{(program $v$ $M$)} we check the version number $v$ and then evaluate the body $M$
+with an empty stack and environment: $\cdot;[] \compute M$.
+
+\subfile{figures/TypedCekMachine.tex}
+
+\subsubsection{Note. Built-in functions in the CEK machine}
+\label{sec:typed-cek-builtins}
+In the CK machine the arguments of built-in functions are always
+closed values, but in the CEK machine arguments are closures.  We
+assume that the mechanism for evaluating built-in functions in the CEK
+machine can deal with closures, and that
+\begin{itemize}
+\item Evaluation of builtin functions always terminates.
+
+\item For closures $C_1, \ldots, C_n$, evaluation of
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\  C_1 \ldots C_n$ \texttt{)}
+fails in the CEK machine if and only if evaluation of
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\ \discharge{C_1} \ldots \discharge{C_n}$ \texttt{)}
+fails according to Figure~\ref{fig:builtins}.
+
+\item For closures $C_1, \ldots, C_n$,
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\  C_1 \ldots C_n$ \texttt{)}
+evaluates to a closure $C$ in the CEK machine if and only if 
+  \texttt{(builtin}\ \textit{bn} $A_1 \ldots A_k\  \discharge{C_1} \ldots \discharge{C_n}$ \texttt{)}
+evaluates to a value $V$ (according to Figure~\ref{fig:builtins})
+such that $\discharge{C} = V$.
```

Comment:
I think this could do with slightly more explanation. I had to stare at it a bit before I realised that it only matters what the environment has for the free variables in the term, and those are pinned down by the constraint here. The environment can have mappings for other variables and they're just irrelevant, so it's fine to leave them unconstrained.

Reply:
Oh hold on,  maybe that's just wrong.  I checked with Roman a few days ago, and builtins are only supposed to return values, not closures. The builtin stuff doesn't currently know anything about closures (the arguments are all discharged in advance), and builtins don't have any way of constructing closures to return to the machine.  I'm not sure what's going to happen in the end.  I think I'll just say that builtins always have to return values for now.

Reply:
I think a polymorphic builtin could return a closure, isn't that one of the issues? i.e. `id` will just pass its argument closure through. Of course, at the moment we fix that by discharging beforehand.

Reply:
> I think a polymorphic builtin could return a closure, isn't that one of the issues? i.e. `id` will just pass its argument closure through. Of course, at the moment we fix that by discharging beforehand.

Exactly. I think we'll just have to wait and see what @effectfully comes up with eventually.  A slight problem is that if builtins have to know about closures then it kind of ties things down to the CK machine a bit too much; we might even have to change the basic operational semantics to talk about closures.

Also, I was talking about builtins returning values earlier.  That's incorrect: we decided some time ago that they should be able to return arbitrary terms.

Reply:
@michaelpj 

> I think a polymorphic builtin could return a closure, isn't that one of the issues? i.e. `id` will just pass its argument closure through.

It could, but the CK machine doesn't know anything about closures and so we'd need to give up on sharing the constant application machinery between the abstract machines, which goes as far as having different implementations of `OpaqueTerm` and probably the whole `KnownType` thing, so it's a big deal. I also really like that `OpaqueTerm` is just a `Term` as that allows to transform that term whatever way I want, which I think we'll need for polymorphic built-in types.

@kwxm 

> Also, I was talking about builtins returning values earlier. That's incorrect: we decided some time ago that they should be able to return arbitrary terms.

Yes, we allow arbitrary terms. There are still some confusing `Value`s sprinkled around the codebase. I think that `Value` should be either removed completely or turned into a `newtype`, otherwise it outdates very quickly.

Reply:
>  I think that `Value` should be either removed completely or turned into a `newtype`, otherwise it outdates very quickly.

(Thumbs up).  It always seemed a bit odd singling out values in the specification but then not distinguishing them in the Haskell type system.

Reply:
I've now updated the document to say that builtins return terms, not closures.